PlantnetBTP status duplicate

log extract

```
19:28:07.377] [ERROR] [default] - [39mTwitterClient | POST statuses/update:{"in_reply_to_status_id":"1253682664839811078","status":"@JeanAlbertTurin Pl@ntnet identifie (à 67.95%) Sempervivum tectorum (fam Crassulaceae) com. Grande joubarbe, Joubarbe des toits, Artichaut bâtard\n\n\nfleur - Aguilar Belem\nhttp://tinyurl.com/ybpqnrzu\n\n#BetPlantnet,#IndentificationDePlantes"} - err:Error: Status is a duplicate.
2020-04-30T19:28:07.377920+00:00 app[web.1]: [31m[2020-04-30 19:28:07.377] [ERROR] [default] - [39m#BetPlantnet,#IndentificationDePlantes | Error: Status is a duplicate.
2020-04-30T19:28:07.378107+00:00 app[web.1]: [33m[2020-04-30 19:28:07.377] [WARN] [default] - [39mBotEngine | plugin response status:500 msg:impossible de répondre au tweet

```

--- a/core/BotEngine.js
+++ b/core/BotEngine.js
@@ -9,7 +9,7 @@ const ENGINE_MIN_INTERVAL_MS=10000;// min each 10 seconds
 
 class BotEngine {
   constructor(newsBotDep, engineMinIntervalMs) {
-    this.logger = log4js.getLogger();
+    this.logger = log4js.getLogger('BotEngine');
     this.logger.setLevel('INFO');
     this.twitterClient = new TwitterClient();
     this.plantnetClient = new MyPlantnetClient();
@@ -25,7 +25,7 @@ class BotEngine {
     engine.plugins.push(filmBTP);
     engine.plugins.push(new PlantnetBTP(engine.twitterClient, engine.plantnetClient));
     engine.defaultPlugin = filmBTP;
-    engine.logInfo("started with " + engine.plugins.length +
+    engine.logger.info("started with " + engine.plugins.length +
        " plugin(s) and minInterval:" + this.intervalMs);
   }
 
@@ -35,7 +35,7 @@ class BotEngine {
     let allowedTs = engine.lastProcess + engine.intervalMs;
     let needToWaitSec = Math.floor((allowedTs - nowMs)/1000);
     if (engine.lastProcess && allowedTs > nowMs) {
-        engine.logInfo(remoteAdd + " | need to wait " + needToWaitSec + " sec" );
+        engine.logger.info(remoteAdd + " | need to wait " + needToWaitSec + " sec" );
         cb({"message": "Demande trop rapprochée, retentez plus tard",
            "status": 429});
         return;
@@ -43,21 +43,21 @@ class BotEngine {
     engine.lastProcess = nowMs;
     let plugin = engine.getPluginByName(pluginName);
     if (!plugin || !plugin.isReady()) {
-        engine.logInfo(remoteAdd + " | no plugin available");
+        engine.logger.info(remoteAdd + " | no plugin available");
         cb({"message": "je suis actuellement en maintenance, retentez plus tard",
             "status": 503});
         return;
     }
-    engine.logInfo(remoteAdd + " | process right now - " + plugin.getPluginTags());
-    engine.newsBot.add("Exécution du plugin - " + plugin.getPluginTags());
+    engine.logger.info(remoteAdd + " | process right now - " + plugin.getName());
+    engine.newsBot.add("Exécution du plugin - " + plugin.getName());
     plugin.process(doSimulate, (err, result) => {
         if(err) {
-          engine.logWarn("plugin response status:" + err.status + " msg:" + err.message);
+          engine.logger.warn("plugin response status:" + err.status + " msg:" + err.message);
           engine.newsBot.add(err.message);
           cb(err);
           return;
         }
-        engine.logInfo("plugin result "+ result.text);
+        engine.logger.info("plugin result "+ result.text);
         engine.newsBot.add(result.html);
         cb(false, result.text);
     });
@@ -77,19 +77,6 @@ class BotEngine {
         return undefined;
     }
     return arr[Math.floor(Math.random() * arr.length)];
-  }
-
-  logWarn(msg) {
-    this.logger.warn("BotEngine | " + msg);
-  }
-  logError(msg) {
-    this.logger.error("BotEngine | " + msg);
-  }
-  logInfo(msg) {
-    this.logger.info("BotEngine | " + msg);
-  }
-  logDebug(msg) {
-    this.logger.debug("BotEngine | " + msg);
   }
 }
 

--- a/core/MyPlantnetClient.js
+++ b/core/MyPlantnetClient.js
@@ -12,7 +12,7 @@ class MyPlantnetClient {
 
   constructor() {
     this.isAvailable = false;
-    this.logger = log4js.getLogger();
+    this.logger = log4js.getLogger('MyPlantnetClient');
     this.logger.setLevel('INFO'); // DEBUG will show api params
 
     try {
@@ -21,9 +21,9 @@ class MyPlantnetClient {
         }
         this.apiKey = process.env.MYPLANTNET_API_PRIVATE_KEY;
         this.isAvailable = true;
-        this.logInfo("available");
+        this.logger.info("available");
     } catch (exception) {
-        this.logError(exception);
+        this.logger.error(exception);
     }
   }
 
@@ -32,7 +32,7 @@ class MyPlantnetClient {
   }
 
   identify(imageUrl, doSimulate, cb) {
-      this.logInfo("identify following image : " + imageUrl);
+      this.logger.info("identify following image : " + imageUrl);
       if (doSimulate) {
         let simulatedAnswer = fs.readFileSync('./core/data/plantNetFrenchResponse.json');
         cb(false, JSON.parse(simulatedAnswer));
@@ -53,7 +53,7 @@ class MyPlantnetClient {
           let errError = err.message;
           let errDetails = err.response.text;
           let errResult = "Pla@ntnet identify error (" + errStatus + ") " + errError;
-          this.logError(errResult + " - details:" + errDetails);
+          this.logger.error(errResult + " - details:" + errDetails);
           cb(errResult);
           return;
         }
@@ -128,16 +128,6 @@ class MyPlantnetClient {
 
   arrayWithContent(arr) {
     return (Array.isArray(arr) && arr.length > 0);
-  }
-
-  logError(msg) {
-    this.logger.error("MyPlantnetClient | " + msg);
-  }
-  logInfo(msg) {
-    this.logger.info("MyPlantnetClient | " + msg);
-  }
-  logDebug(msg) {
-    this.logger.debug("MyPlantnetClient | " + msg);
   }
 }
 

--- a/core/TwitterClient.js
+++ b/core/TwitterClient.js
@@ -7,7 +7,7 @@ const TWEET_MAX_LENGTH=280;
 
 class TwitterClient {
   constructor() {
-    this.logger = log4js.getLogger();
+    this.logger = log4js.getLogger('TwitterClient');
     this.logger.setLevel('INFO'); // DEBUG will show api params
     if (!process.env.APPLICATION_CONSUMER_KEY_HERE ||
     !process.env.APPLICATION_CONSUMER_SECRET_HERE ||
@@ -29,7 +29,7 @@ class TwitterClient {
   // https://developer.twitter.com/en/docs/tweets/post-and-engage/overview
   //   GET statuses/show/:id
   //   https://developer.twitter.com/en/docs/tweets/post-and-engage/api-reference/get-statuses-show-id
-  getTweetDetails(tweetId) {
+  getTweetDetails(tweetId, cb) {
       let params = {
         "id": tweetId,
         "tweet_mode":"extended"
@@ -37,12 +37,35 @@ class TwitterClient {
         // useless "include_entities": true,
       this.twit.get('statuses/show', params, (err, data, response) => {
         if(!err){
-          this.logDebug("GET statuses/show:" + JSON.stringify(params) + " - result:" +
+          this.logger.debug("GET statuses/show:" + JSON.stringify(params) + " - result:" +
             JSON.stringify(data));
-          // cb(false, tweets);
+            cb(false, data);
         } else {
-          this.logError("GET statuses/show:" + params + " - err:" + err);
-          // cb(err);
+          this.logger.error("GET statuses/show:" + params + " - err:" + err);
+          cb(err);
+        }
+      });
+  }
+
+  userTimeline(userName, searchCount, extendedMode, cb) {
+      // doc: https://developer.twitter.com/en/docs/tweets/timelines/api-reference/get-statuses-user_timeline
+      let params = {
+        screen_name: userName,
+        exclude_replies: false,
+        count: searchCount
+      };
+      if (extendedMode) {
+        params.tweet_mode = "extended";
+      }
+
+      this.twit.get('statuses/user_timeline', params, (err, data, response) => {
+        if(!err){
+          let tweets = data;
+          this.logger.debug("GET statuses/user_timeline:" + JSON.stringify(params) + " - result count:" + tweets.length);
+          cb(false, tweets);
+        } else {
+          this.logger.error("GET statuses/user_timeline:" + params + " - err:" + err);
+          cb(err);
         }
       });
   }
@@ -53,9 +76,9 @@ class TwitterClient {
       let params = {
         q: searchQuery,
         count: searchCount,
-        result_type: 'recent',
-        lang: 'fr'
       };
+      params.result_type = 'recent'; // 'mixed';
+      params.lang = 'fr';
       if (extendedMode) {
         params.tweet_mode = "extended";
       }
@@ -63,10 +86,10 @@ class TwitterClient {
       this.twit.get('search/tweets', params, (err, data, response) => {
         if(!err){
           let tweets = data.statuses;
-          this.logDebug("GET search/tweets:" + JSON.stringify(params) + " - result count:" + tweets.length);
+          this.logger.debug("GET search/tweets:" + JSON.stringify(params) + " - result count:" + tweets.length);
           cb(false, tweets);
         } else {
-          this.logError("GET search/tweets:" + params + " - err:" + err);
+          this.logger.error("GET search/tweets:" + params + " - err:" + err);
           cb(err);
         }
       });
@@ -85,9 +108,9 @@ class TwitterClient {
         status: replyStatus
       };
 
-      this.logDebug("Uses params to reply to : " + JSON.stringify(params));
-      this.logInfo(`Plan to reply(${replyLength}) to => ` + this.tweetLinkOf(tweet));
-      this.logInfo(" " + this.tweetInfoOf(tweet));
+      this.logger.debug("Uses params to reply to : " + JSON.stringify(params));
+      this.logger.info(`Plan to reply(${replyLength}) to => ` + this.tweetLinkOf(tweet));
+      this.logger.info(" " + this.tweetInfoOf(tweet));
 
       if (doSimulate) {
           cb(false, {
@@ -106,13 +129,35 @@ class TwitterClient {
       }
       this.twit.post('statuses/update', params, (err, data, res) => {
         if(!err){
-          this.logInfo("POST statuses/update "+ JSON.stringify(params) + " - result" + data);
+          this.logger.info("POST statuses/update "+ JSON.stringify(params) + " - result" + data);
           cb(false, data);
         } else {
-          this.logError("POST statuses/update:" + JSON.stringify(params) + " - err:" + err);
+          this.logger.error("POST statuses/update:" + JSON.stringify(params) + " - err:" + err);
           cb(err);
         }
       });
+  }
+
+  getRecentlyMentionedUsers(userName, count, cb) {
+    let extendedMode = true;
+    this.userTimeline(userName, count, !extendedMode, (err,tweets) => {
+      let mentioned = [];
+      if (err) {
+        cb(err);
+        return;
+      }
+      // this.logger.debug(JSON.stringify(tweets));
+      tweets.forEach((t) => {
+          if (!t.entities || !t.entities.user_mentions) {
+            return;
+          }
+          t.entities.user_mentions.forEach((mention) => {
+              mentioned.push(mention.screen_name);
+          });
+      });
+      cb(false, mentioned);
+      return;
+    });
   }
 
   tweetLinkOf(tweet) {
@@ -165,16 +210,6 @@ class TwitterClient {
     } catch (err) {
       return twitterDate;
     }
-  }
-
-  logError(msg) {
-    this.logger.error("TwitterClient | " + msg);
-  }
-  logInfo(msg) {
-    this.logger.info("TwitterClient | " + msg);
-  }
-  logDebug(msg) {
-    this.logger.debug("TwitterClient | " + msg);
   }
 }
 

--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "botEnTrain",
-  "version": "1.1.1",
+  "version": "1.1.2",
   "description": "Twitter botEnTrain",
   "main": "bet.js",
   "scripts": {

--- a/plugins/FilmBTP.js
+++ b/plugins/FilmBTP.js
@@ -5,16 +5,16 @@ const fs = require('fs');
 class FilmBTP {
   constructor(twitterClient) {
     this.isAvailable = false;
-    this.logger = log4js.getLogger();
+    this.logger = log4js.getLogger('FilmBTP');
     this.logger.setLevel('DEBUG'); // DEBUG will show search results
     this.twitterClient = twitterClient;
     try {
         let behaviorsData = fs.readFileSync('plugins/data/filmBtp.json');
         this.behaviors = JSON.parse(behaviorsData);
         this.isAvailable = true;
-        this.logInfo("available with " + this.behaviors.length + " behaviors");
+        this.logger.info("available with " + this.behaviors.length + " behaviors");
     } catch (exception) {
-        this.logError(exception);
+        this.logger.error(exception);
     }
   }
 
@@ -23,7 +23,7 @@ class FilmBTP {
   }
 
   getPluginTags() {
-    return ["#BetFilm","#RepliquesDeFilms"];
+    return ["#BetFilm","#RepliquesDeFilms"].join(' ');
   }
 
   isReady() {
@@ -34,7 +34,7 @@ class FilmBTP {
     let behavior = this.randomFromArray(this.behaviors);
     let tweets = this.searchTweets(behavior, (err, tweets) => {
         if (err) {
-            this.logError(err);
+            this.logger.error(err);
             cb({ "message": "impossible de chercher des tweets", "status": 500});
             return;
         }
@@ -44,12 +44,12 @@ class FilmBTP {
             debugLog += "\n\t" + this.twitterClient.tweetLinkOf(t) + "\n\t\t" + this.twitterClient.tweetInfoOf(t);
             // debugLog += JSON.stringify(t) ;
         });
-        this.logDebug("search results " + debugLog);
+        this.logger.debug("search results " + debugLog);
         */
         let filteredTweets = tweets.filter((t) => {
             return t.text && t.text.includes(behavior.search);
         });
-        this.logInfo("choose a tweet from results q:" + behavior.search +
+        this.logger.info("choose a tweet from results q:" + behavior.search +
          " results:" + tweets.length +
          " filtered:" + filteredTweets.length);
         let tweetCandidate = this.randomFromArray(filteredTweets);
@@ -60,7 +60,7 @@ class FilmBTP {
         }
         this.replyTweet(doSimulate, tweetCandidate, behavior, (err, replyTweet) => {
             if (err) {
-                this.logError(err);
+                this.logger.error(err);
                 cb({"message": "impossible de répondre au tweet", "status": 500});
                 return;
             }
@@ -99,7 +99,7 @@ class FilmBTP {
                 });
             });
             if (plugin.arrayWithContent(mentioned)) {
-              plugin.logDebug("Exclude already mentioned users => " + mentioned.join(', '));
+              plugin.logger.debug("Exclude already mentioned users => " + mentioned.join(', '));
             }
         }
         let searchQueryNotMe = "\"" + behavior.search + "\"" + noRetweet + notMe;
@@ -128,16 +128,6 @@ class FilmBTP {
 
   arrayWithContent(arr) {
     return (Array.isArray(arr) && arr.length > 0);
-  }
-
-  logDebug(msg) {
-    this.logger.debug(this.getPluginTags() + " | " + msg);
-  }
-  logInfo(msg) {
-    this.logger.info(this.getPluginTags() + " | " + msg);
-  }
-  logError(msg) {
-    this.logger.error(this.getPluginTags() + " | " + msg);
   }
 }
 

--- a/plugins/PlantnetBTP.js
+++ b/plugins/PlantnetBTP.js
@@ -3,20 +3,20 @@ const log4js = require('log4js');
 
 const PLANTNET_MINIMAL_PERCENT = 60;
 const PLANTNET_MINIMAL_RATIO = PLANTNET_MINIMAL_PERCENT / 100;
-const PLANTNET_SIMULATE = false;
+const PLANTNET_SIMULATE = process.env.PLANTNET_SIMULATE || false;
 
 class PlantnetBTP {
   constructor(twitterClient, plantnetClient) {
     this.isAvailable = false;
-    this.logger = log4js.getLogger();
+    this.logger = log4js.getLogger('PlantnetBTP');
     this.logger.setLevel('DEBUG'); // DEBUG will show search results
     this.twitterClient = twitterClient;
     this.plantnetClient = plantnetClient;
     try {
         this.isAvailable = plantnetClient.isReady();
-        this.logInfo(this.isAvailable ? "available" : "not available");
+        this.logger.info(this.isAvailable ? "available" : "not available");
     } catch (exception) {
-        this.logError(exception);
+        this.logger.error(exception);
     }
   }
 
@@ -25,21 +25,31 @@ class PlantnetBTP {
   }
 
   getPluginTags() {
-    return ["#BetPlantnet","#IndentificationDePlantes"];
+    return ["#BetPlantnet","#IndentificationDePlantes"].join(' ');
   }
 
   isReady() {
     return this.isAvailable;
   }
 
+  _debugTweet() {
+      let tweetId = "11223344";
+      this.twitterClient.getTweetDetails(tweetId, (err, data) => {
+          this.logger.error(err);
+          this.logger.info(data);
+      });
+  }
+
   process(doSimulate, cb) {
+    // DEBUG // this._debugTweet(); return;
+
     let question = "(\"quelle est cette plante\" OR \"quelle est cette fleur\")";
     let noArbre = " -arbre";
     let withImage = " filter:media filter:images";
     let plantnetSearch = question + noArbre + withImage;
     this.searchTweets(plantnetSearch, (err, tweets) => {
         if (err) {
-            this.logError(err);
+            this.logger.error(err);
             cb({ "message": "impossible de chercher des tweets", "status": 500});
             return;
         }
@@ -49,7 +59,7 @@ class PlantnetBTP {
             debugLog += "\n\t" + this.twitterClient.tweetLinkOf(t) + "\n\t\t" + this.twitterClient.tweetInfoOf(t);
             // debugLog += JSON.stringify(t) ;
         });
-        this.logDebug("search results " + debugLog);
+        this.logger.debug("search results " + debugLog);
         */
         let tweetCandidate = this.randomFromArray(tweets);
         if (!tweetCandidate) {
@@ -58,7 +68,7 @@ class PlantnetBTP {
             return;
         }
         let candidateImage = this.twitterClient.tweetFirstMediaUrl(tweetCandidate);
-        this.logDebug("tweetCandidate : " +
+        this.logger.debug("tweetCandidate : " +
             this.twitterClient.tweetLinkOf(tweetCandidate) + "\n\t" +
             this.twitterClient.tweetInfoOf(tweetCandidate) + "\n\t" +
             "first media url : " + candidateImage + "\n");
@@ -68,16 +78,16 @@ class PlantnetBTP {
                  "status": 202});
             return;
         }
-        this.logDebug("candidateImage : " + candidateImage);
+        this.logger.debug("candidateImage : " + candidateImage);
 
         this.plantnetClient.identify(candidateImage, PLANTNET_SIMULATE, (err, plantResult) => {
             if (err) {
-                this.logError(err);
+                this.logger.error(err);
                 cb({"message": "impossible d'identifier l'image", "status": 500});
                 return;
             }
 
-            // this.logDebug("plantnetResult : " + JSON.stringify(plantResult));
+            // this.logger.debug("plantnetResult : " + JSON.stringify(plantResult));
 
             let firstScoredResult = this.plantnetClient.hasScoredResult(plantResult, PLANTNET_MINIMAL_RATIO);
             if (!firstScoredResult) {
@@ -100,7 +110,7 @@ class PlantnetBTP {
 
       this.replyTweet(doSimulate, tweetCandidate, replyMessage, (err, replyTweet) => {
           if (err) {
-              this.logError(err);
+              this.logger.error(err);
               cb({"message": "impossible de répondre au tweet", "status": 500});
               return;
           }
@@ -125,7 +135,7 @@ class PlantnetBTP {
       " mais cela n'a pas donné de résultat concluant (score>" + PLANTNET_MINIMAL_PERCENT + "%).\n:(\n";
       this.replyTweet(doSimulate, tweetCandidate, replyMessage, (err, replyTweet) => {
           if (err) {
-              this.logError(err);
+              this.logger.error(err);
               cb({"message": "impossible de répondre au tweet", "status": 500});
               return;
           }
@@ -147,30 +157,23 @@ class PlantnetBTP {
   searchTweets(plantnetSearch, cb) {
     let plugin = this;
     let noRetweet = " -filter:retweets";
-    let fromMe = " from:botEnTrain1";
     let notMe = " -from:botEnTrain1";
-    let searchQueryFromMe = this.getPluginTags()[0] + noRetweet + fromMe;
     let extendedMode = true;
-    plugin.twitterClient.search(searchQueryFromMe, 200, !extendedMode, (err,tweets) => {
-        let mentioned = [];
-        if (!err) {// get already mentioned users
-            tweets.forEach((t) => {
-                if (!t.entities || !t.entities.user_mentions) {
-                  return;
-                }
-                t.entities.user_mentions.forEach((mention) => {
-                    mentioned.push(mention.screen_name);
-                });
-            });
-            if (plugin.arrayWithContent(mentioned)) {
-              plugin.logDebug("Exclude already mentioned users => " + mentioned.join(', '));
-            }
+    plugin.logger.debug("get already mentioned users");
+    plugin.twitterClient.getRecentlyMentionedUsers("botEnTrain1", 200, (err, mentionedUsers) => {
+        if (err) {
+            plugin.logger.warn("Unable to search mentioned users " + JSON.stringify(err));
+            cb(err);
+            return;
+        }
+        if (plugin.arrayWithContent(mentionedUsers)) {
+          plugin.logger.debug("Exclude already mentioned users => " + mentionedUsers.join(', '));
         }
         let searchQueryNotMe = plantnetSearch + noRetweet + notMe;
-        mentioned.forEach((mentionedUser) => {// exclude alreadyMentioned
+        mentionedUsers.forEach((mentionedUser) => {// exclude alreadyMentioned
             searchQueryNotMe += " -from:" + mentionedUser;
         });
-        plugin.twitterClient.search(searchQueryNotMe, 20, extendedMode, cb);
+        plugin.twitterClient.search(searchQueryNotMe, 50, extendedMode, cb);
     });
   }
 
@@ -188,16 +191,6 @@ class PlantnetBTP {
 
   arrayWithContent(arr) {
     return (Array.isArray(arr) && arr.length > 0);
-  }
-
-  logDebug(msg) {
-    this.logger.debug(this.getPluginTags() + " | " + msg);
-  }
-  logInfo(msg) {
-    this.logger.info(this.getPluginTags() + " | " + msg);
-  }
-  logError(msg) {
-    this.logger.error(this.getPluginTags() + " | " + msg);
   }
 }
 

--- a/plugins/data/filmBtp.json
+++ b/plugins/data/filmBtp.json
@@ -68,7 +68,7 @@
   },
   {
     "indexIs": 11,
-    "search": "l'heure de se réveiller",
+    "search": "il est l'or",
     "reply": "Monseignor, il est l'or, l'or de se réveiller",
     "link": "https://youtu.be/_sJYrReI3h8",
     "credits": "Yves Montand (Blaze) et Louis de Funès (don Salluste) dans La folie des Grandeurs (1971)"

--- a/scripts/env.template.sh
+++ b/scripts/env.template.sh
@@ -9,4 +9,6 @@ export ACCESS_TOKEN_SECRET_HERE=
 export TOKEN_SIMULATION=
 export TOKEN_ACTION=
 
+# PLANTNET_SIMULATE=true will simulate Pl@ntNet API result
+export PLANTNET_SIMULATE=true
 export MYPLANTNET_API_PRIVATE_KEY=

